cdap-ui/app/directives/timeline/timeline.js

```diff
@@ -268,45 +287,137 @@ function link (scope, element) {
 
   scope.updateSlider = updateSlider;
 
-  var generateEventCircles = function (){
-    let circleClass;
+  const generateEventCircles = () => {
+
+    //Clears the tooltip regularly
+    d3.selectAll('.circle-tooltip').remove();
+
+    circleTooltip = d3.select('body').append('div')
+      .attr('class', 'circle-tooltip')
+      .style('opacity', 0);
+
+    timelineStack = {};
+    let errorMap = {};
+    let warningMap = {};
+    timelineData = scope.metadata;
+
+    //Subroutine
+    const mapEvents = (type, data) => {
+      let typeMap;
+      if(type === 'warn'){
+        typeMap = warningMap;
+      } else if(type === 'error'){
+        typeMap = errorMap;
+      }
+
+      for(let k = 0; k < data.length; k++){
+        let currentItem = data[k];
+        let xVal = Math.floor(xScale(currentItem.time * 1000));
+        typeMap[xVal] = currentItem.value;
+      }
+      return typeMap;
+    };
 
     if(timelineData.qid.series.length > 0){
       for(let i = 0; i < timelineData.qid.series.length; i++){
-
         switch(timelineData.qid.series[i].metricName){
           case 'system.app.log.error':
-            circleClass = 'red-circle';
+            errorMap = mapEvents('error', timelineData.qid.series[i].data);
             break;
           case 'system.app.log.warn':
-            circleClass = 'yellow-circle';
+            warningMap = mapEvents('warn', timelineData.qid.series[i].data);
             break;
           default:
-            circleClass = 'other-circle';
             break;
         }
+      }
+    }
 
-        for(let j = 0; j < timelineData.qid.series[i].data.length; j++){
-          let currentItem = timelineData.qid.series[i].data[j];
-          let xVal = Math.floor(xScale(currentItem.time));
-          let numEvents = currentItem.value;
-
-          if(timelineStack[xVal] === undefined) {
-            timelineStack[xVal] = 0;
-          }
+    const circleTooltipHoverOn = function(position, errors, warnings) {
+      let date = xScale.invert(position).toString();
+      date = date.substring(0, date.length - 14);
```

Comment:
why 14? you should put a comment on that. Also if you are offsetting by timezone, what would happen for people using this in other timezone? having a static value seems incorrect